crates/adapters/src/integrated/delta_table/input.rs

```diff
@@ -0,0 +1,688 @@
+use crate::catalog::{ArrowStream, InputCollectionHandle};
+use crate::controller::{ControllerInner, EndpointId};
+use crate::integrated::delta_table::delta_input_serde_config;
+use crate::transport::{InputEndpoint, IntegratedInputEndpoint, Step};
+use crate::{
+    ControllerError, InputConsumer, InputReader, ParseError, PipelineState, RecordFormat,
+    TransportInputEndpoint,
+};
+use anyhow::{anyhow, Result as AnyResult};
+use dbsp::InputHandle;
+use deltalake::datafusion::common::Column;
+use deltalake::datafusion::dataframe::DataFrame;
+use deltalake::datafusion::execution::context::SQLOptions;
+use deltalake::datafusion::prelude::{Expr, ParquetReadOptions, SessionContext};
+use deltalake::datafusion::sql::sqlparser::dialect::GenericDialect;
+use deltalake::datafusion::sql::sqlparser::keywords::Keyword::SQL;
+use deltalake::datafusion::sql::sqlparser::parser::Parser;
+use deltalake::datafusion::sql::sqlparser::test_utils::table;
+use deltalake::delta_datafusion::cdf::FileAction;
+use deltalake::kernel::Action;
+use deltalake::kernel::Error::Parse;
+use deltalake::operations::create::CreateBuilder;
+use deltalake::protocol::SaveMode;
+use deltalake::table::PeekCommit;
+use deltalake::{datafusion, DeltaTable, DeltaTableBuilder, Path};
+use env_logger::builder;
+use futures_util::StreamExt;
+use log::{debug, error, info, trace};
+use pipeline_types::config::{InputEndpointConfig, TransportConfigVariant};
+use pipeline_types::format::json::JsonFlavor;
+use pipeline_types::program_schema::Relation;
+use pipeline_types::transport::delta_table::{DeltaTableIngestMode, DeltaTableReaderConfig};
+use pipeline_types::transport::s3::{ReadStrategy, S3InputConfig};
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::fmt::format;
+use std::str::FromStr;
+use std::sync::{Arc, Weak};
+use std::time::Duration;
+use tokio::select;
+use tokio::sync::mpsc;
+use tokio::sync::watch::{channel, Receiver, Sender};
+use tokio::time::sleep;
+use url::Url;
+use utoipa::ToSchema;
+
+/// Polling interval when following a delta table.
+const POLL_INTERVAL: Duration = Duration::from_millis(1000);
+
+/// Polling delay before retrying an unsuccessful read from a delta log.
+const RETRY_INTERVAL: Duration = Duration::from_millis(2000);
+
+/// Integrated input connector that reads from a delta table.
+pub struct DeltaTableInputEndpoint {
+    inner: Arc<DeltaTableInputEndpointInner>,
+}
+
+impl DeltaTableInputEndpoint {
+    pub fn new(
+        endpoint_id: EndpointId,
+        endpoint_name: &str,
+        config: &DeltaTableReaderConfig,
+        controller: Weak<ControllerInner>,
+    ) -> Self {
+        // Register url handlers, so URL's like `s3://...` are recognized.
+        // I hope it is harmless to do this every time a DeltaTablWriter is created.
+        deltalake::aws::register_handlers(None);
+        deltalake::azure::register_handlers(None);
+        deltalake::gcp::register_handlers(None);
```

Comment:
The deltalake code is really unclear about that (and there are no docs), so doing this just once would probably be safer.

Reply:
I implemented at most once initialization to be on the safe side.